### PR TITLE
add naming tutorial

### PR DESCRIPTION
**Pull request by piiswrong**
## Description ##
(Brief description on what this PR is about)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: docs/tutorials/gluon/naming.md [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+
+Now let's instantiate our neural net.
+
+- Note that `model0.dense0` is named as `model0_dense0_` instead of `dense0_`.
+
+- Also note that although we specified `mydense_` as prefix for `model.mydense`, its parent's prefix is automatically prepended to generate the prefix `model0_mydense_`.
+
+
+```python
+model0 = Model()
```
[the remark this answers is not in the document]
Comment (piiswrong) [reply docs/tutorials/gluon/naming.md]: I think model0 is fine. It matches the prefix

File: docs/tutorials/gluon/naming.md [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+
+```python
+with alexnet.name_scope():
+    alexnet.output = gluon.nn.Dense(100)
+alexnet.output.initialize()
+print(alexnet.output, alexnet.output.prefix)
+```
+
+    Dense(None -> 100, linear) alexnet0_dense3_
+
```
Comment (ThomasDelteil) [inline docs/tutorials/gluon/naming.md]: can you please add `<!-- INSERT SOURCE DOWNLOAD BUTTONS -->` to enable the notebook download of your tutorial?

File: docs/tutorials/gluon/naming.md [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+**Nevertheless, you are still recommended to manually specify the prefix of the top level Block (i.e. `model = Model(prefix='mymodel_')`) to avoid any potential problem during saving and loading.**
+
+## Replacing Blocks from networks and fine-tuning
+
+Sometimes you may want to load a pretrained model, and replace certain Blocks in it for fine-tuning.
+
+For example, the alexnet in model zoo has 1000 output dimensions, but maybe you only have 100 classes in your application.
+
+To see how to do this, we first load an pretrained alexnet.
```
Comment (ThomasDelteil) [inline docs/tutorials/gluon/naming.md]: a pretrained*

File: docs/tutorials/gluon/naming.md [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+
+
+To change the output to 100 dimension, we replace it with a new block.
+
+- Note that it's important to do this in alexnet's name_scope, otherwise you will have unmatching names when you try to save and load your model.
+
+
+```python
+with alexnet.name_scope():
+    alexnet.output = gluon.nn.Dense(100)
```
Comment (ThomasDelteil) [inline docs/tutorials/gluon/naming.md]: Maybe it is worth putting here that this works only because the models in model zoo are inheriting `HybridBlock` and have been built in a way so that their final output layer is associated with the `.output` property.

I can see people building their own network using`HybridSequential` for example and trying to update the output layer by calling `output` on their network.

**Comment by ThomasDelteil** [comment]
Great tutorial, it was much needed and clarifies what is going on under the hood with the naming scopes 👍

File: docs/tutorials/gluon/naming.md [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+      Parameter dense0_weight (shape=(100, 0), dtype=<type 'numpy.float32'>)
+      Parameter dense0_bias (shape=(100,), dtype=<type 'numpy.float32'>)
+    )
+
+
+## Name scopes
+
+To manage the names of nested Blocks, each Block has a `name_scope` attached to it. All Blocks created within a name scope will have its parent Block's prefix prepended to its name.
+
+Let's demonstrate this by first define a simple neural net:
```
Comment (aaronmarkham) [inline docs/tutorials/gluon/naming.md]: defining

File: docs/tutorials/gluon/naming.md [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+- Note that `model0.dense0` is named as `model0_dense0_` instead of `dense0_`.
+
+- Also note that although we specified `mydense_` as prefix for `model.mydense`, its parent's prefix is automatically prepended to generate the prefix `model0_mydense_`.
+
+
+```python
+model0 = Model()
+model0.initialize()
+model0(mx.nd.zeros((1, 20)))
+print(model0.prefix, model0.dense0.prefix, model0.dense1.prefix, model0.mydense.prefix)
```
Comment (aaronmarkham) [inline docs/tutorials/gluon/naming.md]: This print all runs together, and... maybe a I'm a little dense ;) but I'm not quite catching the point. Plus I have a hard time breaking it up when the Jupyter font uses an 'l' that looks like a '1', so I see mode10, not model0, and run together it's incomprehensible.

File: docs/tutorials/gluon/naming.md [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+
+```python
+model1 = Model()
+print(model1.prefix, model1.dense0.prefix, model1.dense1.prefix, model1.mydense.prefix)
+```
+
+    model1_ model1_dense0_ model1_dense1_ model1_mydense_
+
+
+**It is recommended that you manually specify prefix for the top level Block (i.e. `model = Model(prefix='mymodel_')`) to avoid potential confusions in naming**
```
Comment (aaronmarkham) [inline docs/tutorials/gluon/naming.md]: specify **a** prefix

Comment (aaronmarkham) [reply docs/tutorials/gluon/naming.md]: add a period at the end

File: docs/tutorials/gluon/naming.md [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+
+The same principle also applies to container blocks like Sequantial. `name_scope` can be used inside `__init__` as well as out side of `__init__`:
```
Comment (aaronmarkham) [inline docs/tutorials/gluon/naming.md]: spelling... Sequential

File: docs/tutorials/gluon/naming.md [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+from mxnet import gluon
+```
+
+## Naming Blocks
+
+When creating a block, you can assign a prefix to it:
+
+
+```python
+mydense = gluon.nn.Dense(100, prefix='mydense_')
```
Comment (aaronmarkham) [inline docs/tutorials/gluon/naming.md]: You can also choose to not use an underscore and name it what you want. Maybe explain why you want to use an underscore.

File: docs/tutorials/gluon/naming.md [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+
+The same principle also applies to container blocks like Sequantial. `name_scope` can be used inside `__init__` as well as out side of `__init__`:
+
+
+```python
+net = gluon.nn.Sequential()
+with net.name_scope():
+    net.add(gluon.nn.Dense(20))
+    net.add(gluon.nn.Dense(20))
+print(net.prefix, net[0].prefix, net[1].prefix)
```
Comment (aaronmarkham) [inline docs/tutorials/gluon/naming.md]: I tried to mess with the names with regard to the underscore to no ill effect. So what's with it? I used this and it seems fine.
```
net = gluon.nn.Sequential()
with net.name_scope():
    net.add(gluon.nn.Dense(20, prefix='berries'))
    net.add(gluon.nn.Dense(20, prefix='nuts'))
print(net.prefix, net[0].prefix, net[1].prefix)
```
```
sequential4_ sequential4_berries sequential4_nuts
```

File: docs/tutorials/gluon/naming.md (hunk 7 above)
Comment (aaronmarkham) [reply docs/tutorials/gluon/naming.md]: And why not show that code sample if it's so important?

File: docs/tutorials/gluon/naming.md [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+    net.add(gluon.model_zoo.vision.alexnet(pretrained=True))
+print(net.prefix, net[0].prefix, net[1].prefix)
+```
+
+    sequential1_ sequential1_alexnet0_ sequential1_alexnet1_
+
+
+## Saving and loading
+
+Because model0 and model1 have different prefixes, their Parameters also have different names:
```
Comment (aaronmarkham) [inline docs/tutorials/gluon/naming.md]: lower case parameters

File: docs/tutorials/gluon/naming.md [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+
+
+```python
+print(model0.collect_params(), '\n')
+print(model1.collect_params())
+```
+
+    model0_ (
+      Parameter model0_dense0_weight (shape=(20L, 20L), dtype=<type 'numpy.float32'>)
```
Comment (aaronmarkham) [inline docs/tutorials/gluon/naming.md]: Aha! The _ matters with weights and bias and makes thing more readable. Might be cool to demo this kind of issue earlier, and it's good for people to see it happen.
```
model8_ (
  Parameter model8_berriesweight (shape=(20, 20), dtype=<class 'numpy.float32'>)
  Parameter model8_berriesbias (shape=(20,), dtype=<class 'numpy.float32'>)
  Parameter model8_dense0_weight (shape=(20, 20), dtype=<class 'numpy.float32'>)
  Parameter model8_dense0_bias (shape=(20,), dtype=<class 'numpy.float32'>)
  Parameter model8_nutsweight (shape=(20, 20), dtype=<class 'numpy.float32'>)
  Parameter model8_nutsbias (shape=(20,), dtype=<class 'numpy.float32'>)
) 
```

File: docs/tutorials/gluon/naming.md [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+      Parameter model1_dense0_weight (shape=(20, 0), dtype=<type 'numpy.float32'>)
+      Parameter model1_dense0_bias (shape=(20,), dtype=<type 'numpy.float32'>)
+      Parameter model1_dense1_weight (shape=(20, 0), dtype=<type 'numpy.float32'>)
+      Parameter model1_dense1_bias (shape=(20,), dtype=<type 'numpy.float32'>)
+      Parameter model1_mydense_weight (shape=(20, 0), dtype=<type 'numpy.float32'>)
+      Parameter model1_mydense_bias (shape=(20,), dtype=<type 'numpy.float32'>)
+    )
+
+
+As a result if you try to save parameters from model0 and load it with model1, you'll get an error due to unmatching names:
```
Comment (aaronmarkham) [inline docs/tutorials/gluon/naming.md]: result,

File: docs/tutorials/gluon/naming.md [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+
+
+```python
+model0.collect_params().save('model.params')
+try:
+    model1.collect_params().load('model.params', mx.cpu())
+except Exception, e:
```
Comment (aaronmarkham) [inline docs/tutorials/gluon/naming.md]: Not python 3 compatible syntax , --> as

File: docs/tutorials/gluon/naming.md [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+    ['dense0_bias', 'mydense_weight', 'mydense_bias', 'dense1_bias', 'dense1_weight', 'dense0_weight']
+
+
+**Nevertheless, you are still recommended to manually specify the prefix of the top level Block (i.e. `model = Model(prefix='mymodel_')`) to avoid any potential problem during saving and loading.**
+
+## Replacing Blocks from networks and fine-tuning
+
+Sometimes you may want to load a pretrained model, and replace certain Blocks in it for fine-tuning.
+
+For example, the alexnet in model zoo has 1000 output dimensions, but maybe you only have 100 classes in your application.
```
Comment (aaronmarkham) [inline docs/tutorials/gluon/naming.md]: the AlexNet model from the model zoo (add link?)

1,000

File: docs/tutorials/gluon/naming.md (hunk 3 above)
Comment (aaronmarkham) [reply docs/tutorials/gluon/naming.md]: maybe `alexnet` if you want to refer to it now as an object instead of a model from the zoo, where it would be AlexNet.

File: docs/tutorials/gluon/naming.md [hunk 16]
```diff
@@ -0,0 +1,236 @@
...
+alexnet = gluon.model_zoo.vision.alexnet(pretrained=True)
+print(alexnet.output, alexnet.output.prefix)
+```
+
+    Dense(4096 -> 1000, linear) alexnet0_dense2_
+
+
+To change the output to 100 dimension, we replace it with a new block.
+
+- Note that it's important to do this in alexnet's name_scope, otherwise you will have unmatching names when you try to save and load your model.
```
Comment (aaronmarkham) [inline docs/tutorials/gluon/naming.md]: I ran into similar issues and just blanked out the prefix. Here's an example that I used.
https://gist.github.com/aaronmarkham/1017664fe683596c614961112a867145#gistcomment-2336981
Can you explain a bit more? This is all really helpful, and I wonder how you'd explain what I ran into and better ways to fix it.

Like how I used `ignore_extra` to [get around the errors](https://gist.github.com/aaronmarkham/1017664fe683596c614961112a867145#file-gluon-export-model-L30) but was still worried that I hadn't really loaded the model properly.

File: docs/tutorials/gluon/naming.md (hunk 10 above)
Comment (piiswrong) [reply docs/tutorials/gluon/naming.md]: to make it look nicer

File: docs/tutorials/gluon/naming.md (hunk 16 above)
Comment (piiswrong) [reply docs/tutorials/gluon/naming.md]: make a fix so that it won't be a problem anymore

File: python/mxnet/gluon/block.py [hunk 17]
```diff
@@ -187,17 +187,17 @@ def __setattr__(self, name, value):
...
-                        self._children[i] = value
-            elif isinstance(value, Block):
-                self.register_child(value)
-        elif isinstance(value, Block):
-            self.register_child(value)
+                                'is not allowed.'.format(
+                                    name=name, type1=type(existing), type2=type(value)))
+
+        if isinstance(value, Block):
+            self.register_child(value, name)
```
[the remark this answers is not in the document]
Comment (piiswrong) [reply python/mxnet/gluon/block.py]: _children is OrderedDict now

[the remark this answers is not in the document]
Comment (szha) [reply python/mxnet/gluon/block.py]: Just noticed. This means it's finally possible to print the model `repr` in order.

**Comment by zhanghang1989** [comment]
Still have problem save and load. https://github.com/apache/incubator-mxnet/issues/10544
